README
What is it?
===========

This plugin enables the embedding of arbitrary application windows into the Xfce
panel. The window is resized into the panel space available, and the associated
program can be automatically launched if it is not open.

Example uses include embedding an instant messaging buddy list, a mail client's
new mail ticker, a simple media application, or a fancy clock or timer.
Combining with Xfce's ability to auto-hide panels can make this very convenient.


Installation
============

The file 'INSTALL' contains generic installation instructions. For more
detailed information, visit the Embed plugin website at
http://goodies.xfce.org/projects/panel-plugins/xfce4-embed-plugin


Sample Usage
============

The following are some sample configurations.
Any selection criteria not mentioned are assumed to be blank.
Any specified labels are suggested, and can certainly be modified to taste.
Note that in general, you will have to click the label to get keyboard focus.

Embed Pidgin's buddy list:
    Launch command (optional): pidgin
    Process name: pidgin
    Window title: ^Buddy List$

Embed gvim to edit a notes file:
    Launch command: gvim --socketid %s ~/path/to/note/file
    Label format: Notes

Embed a terminal:
    Launch command: xterm -into %s
    Label format: %t


How to report bugs?
===================

Bugs should be reported to the Xfce bug tracking system:
    http://bugzilla.xfce.org/enter_bug.cgi?product=xfce4-embed-plugin
You will need to create an account for yourself.


NEWS
1.6.0
=====
- Add support for drag and drop
- Now translated into 36 languages. Thanks, translators!

1.4.1
=====
- Replace AM_CONFIG_HEADER with AC_CONFIG_HEADERS in configure

1.4.0
=====
- Added optional handle
- Added "close" menu item
- Updated translations

1.2.0
=====
- Improved painting of the socket background
- Made the socket background clickable
- Translations: Arabic, French, Galician, Portugese, Portugese, Spanish, Uyghur
- Updated license text

1.0.0
=====
- Plugin is now compiled as a module
- Now translated into 15 languages

0.8.0
=====
- Initial release, lacking translations


autogen.sh
#!/bin/sh
#
# Copyright (c) 2002-20011 The Xfce development team. All rights reserved.
#
# This library is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the Free
# Software Foundation; either version 2 of the License, or (at your option)
# any later version.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
#
# Written for Xfce by Benedikt Meurer <benny@xfce.org>.
#

(type xdt-autogen) >/dev/null 2>&1 || {
  cat >&2 <<EOF
autogen.sh: You don't seem to have the Xfce development tools installed on
            your system, which are required to build this software.
            Please install the xfce4-dev-tools package first, it is available
            from http://www.xfce.org/.
EOF
  exit 1
}

XDT_AUTOGEN_REQUIRED_VERSION="4.7.2" \
exec xdt-autogen $@


panel-plugin/ewmh.c
/*  xfce4-embed-plugin - Embed arbitrary windows into the Xfce panel
 *
 *  Copyright (C) 2012 David Schneider <dnschneid@gmail.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif
#ifdef HAVE_STRING_H
#include <string.h>
#endif

#include <gtk/gtk.h>
#include <X11/Xlib.h>
#include <X11/Xatom.h>
#include <libxfce4util/libxfce4util.h>
#include <libxfce4panel/xfce-panel-plugin.h>

#include "ewmh.h"

/* EWMH functions adapted from wmctrl project. */

#define MAX_PROPERTY_VALUE_LEN 4096


static gchar *get_property (Display *disp, Window win, Atom xa_prop_type,
                            gchar *prop_name, gulong *size)
{
    Atom xa_prop_name;
    Atom xa_ret_type;
    gint ret_format;
    gulong ret_nitems;
    gulong ret_bytes_after;
    gulong tmp_size;
    guchar *ret_prop;
    gchar *ret;
    
    xa_prop_name = XInternAtom(disp, prop_name, False);
    
    /* MAX_PROPERTY_VALUE_LEN / 4 explanation (XGetWindowProperty manpage):
     *
     * long_length = Specifies the length in 32-bit multiples of the
     *               data to be retrieved.
     *
     * NOTE:  see 
     * http://mail.gnome.org/archives/wm-spec-list/2003-March/msg00067.html
     * In particular:
     *
     * 	When the X window system was ported to 64-bit architectures, a
     * rather peculiar design decision was made. 32-bit quantities such
     * as Window IDs, atoms, etc, were kept as longs in the client side
     * APIs, even when long was changed to 64 bits.
     *
     */
    if (XGetWindowProperty(disp, win, xa_prop_name, 0,
            MAX_PROPERTY_VALUE_LEN / 4, False,
            xa_prop_type, &xa_ret_type, &ret_format,     
            &ret_nitems, &ret_bytes_after, &ret_prop) != Success) {
        DBG("Cannot get %s property.", prop_name);
        return NULL;
    }
  
    if (xa_ret_type != xa_prop_type) {
        DBG("Invalid type of %s property.", prop_name);
        XFree(ret_prop);
        return NULL;
    }

    /* null terminate the result to make string handling easier */
    tmp_size = (ret_format / 8) * ret_nitems;
    /* Correct 64 Architecture implementation of 32 bit data */
    if(ret_format==32) tmp_size *= sizeof(glong)/4;
    ret = g_malloc(tmp_size + 1);
    memcpy(ret, ret_prop, tmp_size);
    ret[tmp_size] = '\0';

    if (size) {
        *size = tmp_size;
    }
    
    XFree(ret_prop);
    return ret;
}

static void client_msg(Display *disp, Window win, gchar *msg, gulong data0) {
    XEvent event;
    glong mask = SubstructureRedirectMask | SubstructureNotifyMask;

    event.xclient.type = ClientMessage;
    event.xclient.serial = 0;
    event.xclient.send_event = True;
    event.xclient.message_type = XInternAtom(disp, msg, False);
    event.xclient.window = win;
    event.xclient.format = 32;
    event.xclient.data.l[0] = data0;
    event.xclient.data.l[1] = 0;
    event.xclient.data.l[2] = 0;
    event.xclient.data.l[3] = 0;
    event.xclient.data.l[4] = 0;

    if (!XSendEvent(disp, DefaultRootWindow(disp), False, mask, &event))
        DBG("Cannot send %s event.", msg);
    XSync (disp, False);
}

Window *get_client_list (Display *disp, gulong *size)
{
    Window *client_list;

    if ((client_list = (Window *)get_property(disp, DefaultRootWindow(disp), 
                    XA_WINDOW, "_NET_CLIENT_LIST", size)) == NULL) {
        if ((client_list = (Window *)get_property(disp, DefaultRootWindow(disp),
                        XA_CARDINAL, "_WIN_CLIENT_LIST", size)) == NULL) {
            DBG("Cannot get client list properties "
                  "(_NET_CLIENT_LIST nor _WIN_CLIENT_LIST).");
            return NULL;
        }
    }

    return client_list;
}


gchar *get_client_proc (Display *disp, Window win)
{
    gulong *pid;
    gchar  proc[25];
    gchar *contents;
    gsize  length;
    gchar *start;
    gchar *end;
    gchar *procname;

    if ((pid = (gulong *)get_property(disp, win, XA_CARDINAL,
                                      "_NET_WM_PID", NULL)) == NULL)
        return NULL;

    procname = NULL;

    /* First attempt the full command line */
    g_snprintf (proc, 25, "/proc/%lu/cmdline", *pid);
    if ((g_file_get_contents (proc, &contents, &length, NULL))) {
        if (length) {
            contents[length-1] = '\0';
            start = strrchr (contents, '/');
            start = start ? start+1 : contents;
            procname = g_strdup (start);
        }
        g_free (contents);
    }
    if (procname == NULL) {
        /* Otherwise, grab the potentially-truncated version from stat */
        g_snprintf (proc, 25, "/proc/%lu/stat", *pid);
        if (g_file_get_contents (proc, &contents, &length, NULL)) {
            if (length) {
                contents[length-1] = '\0';
                start = strchr (contents, '(');
                if (start) {
                    end = strchr (start, ')');
                    if (end) {
                        *end = '\0';
                        procname = g_strdup (start+1);
                    }
                }
            }
            g_free (contents);
        }
    }

    g_free (pid);
    return procname;
}


gchar *get_window_title (Display *disp, Window win)
{
    gchar *title;
    gchar *wm_name;

    wm_name = get_property(disp, win, XInternAtom(disp, "UTF8_STRING", False),
                           "_NET_WM_NAME", NULL);

    if (wm_name) {
        if ((title = g_locale_from_utf8(wm_name, -1, NULL, NULL, NULL))) {
            g_free(wm_name);
            return title;
        }
        return wm_name;
    } else {
        return get_property(disp, win, XA_STRING, "WM_NAME", NULL);
    }
}


gchar *get_window_class (Display *disp, Window win)
{
    gchar *wm_class;
    unsigned long size;

    wm_class = get_property(disp, win, XA_STRING, "WM_CLASS", &size);
    if (wm_class) {
        gchar *p_0 = strchr(wm_class, '\0');
        if (wm_class + size - 1 > p_0) {
            *(p_0) = '.';
        }
    }

    return wm_class;
}


void get_window_size (Display *disp, Window win, gint *width, gint *height)
{
    Window root;
    gint relx, rely;
    guint bw, depth;
    XGetGeometry (disp, win, &root, &relx, &rely,
                  (guint *)width, (guint *)height, &bw, &depth);
}


void make_window_toplevel (Display *disp, Window win, gint width, gint height)
{
    reparent_window (disp, win, DefaultRootWindow (disp), width, height);
}

void reparent_window (Display *disp, Window win, Window parent,
                      gint width, gint height)
{
    XReparentWindow (disp, win, parent, 0, 0);
    resize_window (disp, win, width, height);
}

void resize_window (Display *disp, Window win, gint width, gint height)
{
    if (width > 0 && height > 0)
        XResizeWindow (disp, win, (guint)width, (guint)height);
    XSync (disp, False);
}

void show_window (Display *disp, Window win)
{
    gulong *cur_desktop = NULL;
    Window root = DefaultRootWindow(disp);

    if (! (cur_desktop = (unsigned long *)get_property(disp, root,
            XA_CARDINAL, "_NET_CURRENT_DESKTOP", NULL))) {
        if (! (cur_desktop = (unsigned long *)get_property(disp, root,
                XA_CARDINAL, "_WIN_WORKSPACE", NULL))) {
            DBG("Cannot get current desktop properties "
                "(_NET_CURRENT_DESKTOP nor _WIN_WORKSPACE).");
            return;
        }
    }

    client_msg(disp, win, "_NET_WM_DESKTOP", *cur_desktop);
    g_free(cur_desktop);
}

void focus_window (Display *disp, Window win)
{
    XSetInputFocus (disp, win, RevertToNone, CurrentTime);
    XFlush (disp);
}

void close_window (Display *disp, Window win)
{
    XEvent event;
    event.xclient.type = ClientMessage;
    event.xclient.serial = 0;
    event.xclient.send_event = False;
    event.xclient.message_type = XInternAtom(disp, "WM_PROTOCOLS", False);
    event.xclient.window = win;
    event.xclient.format = 32;
    event.xclient.data.l[0] = XInternAtom(disp, "WM_DELETE_WINDOW", False);
    event.xclient.data.l[1] = 0;
    event.xclient.data.l[2] = 0;
    event.xclient.data.l[3] = 0;
    event.xclient.data.l[4] = 0;
    if (!XSendEvent(disp, win, False, NoEventMask, &event))
        DBG("Cannot send WM_DELETE_WINDOW event.");
    XSync (disp, False);
}


panel-plugin/ewmh.h
/*  xfce4-embed-plugin - Embed arbitrary windows into the Xfce panel
 *
 *  Copyright (C) 2012 David Schneider <dnschneid@gmail.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef __EWMH_H__
#define __EWMH_H__

G_BEGIN_DECLS

Window *get_client_list  (Display *disp, gulong *size);
gchar  *get_client_proc  (Display *disp, Window win);
gchar  *get_window_title (Display *disp, Window win);
gchar  *get_window_class (Display *disp, Window win);
void    get_window_size  (Display *disp, Window win, gint *width, gint *height);

void make_window_toplevel (Display *disp, Window win, gint width, gint height);
void reparent_window (Display *disp, Window win, Window parent,
		      gint width, gint height);
void resize_window (Display *disp, Window win, gint width, gint height);
void show_window (Display *disp, Window win);
void focus_window (Display *disp, Window win);
void close_window (Display *disp, Window win);

G_END_DECLS

#endif /* !__EWMH_H__ */



panel-plugin/embed.h
/*  xfce4-embed-plugin - Embed arbitrary windows into the Xfce panel
 *
 *  Copyright (C) 2012 David Schneider <dnschneid@gmail.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef __EMBED_H__
#define __EMBED_H__

G_BEGIN_DECLS

/* plugin structure */
typedef struct
{
    XfcePanelPlugin *plugin;

    /* panel widgets */
    GtkWidget       *hvbox;
    GtkWidget       *handle;
    GtkWidget       *label;
    GtkWidget       *socket;
    GtkWidget       *embed_menu;
    GtkWidget       *focus_menu;
    GtkWidget       *popout_menu;
    GtkWidget       *close_menu;

    /* panel data */
    gboolean         has_plug;
    GdkNativeWindow  plug;
    GdkWindow       *plug_window;
    gint             plug_width;
    gint             plug_height;
    gboolean         plug_is_gtkplug;

    Display         *disp;

    guint            search_timer;
    guint            search_idle;
    gboolean         disable_search;
    gboolean         monitor_saw_net_client_list;
    gboolean         monitor_saw_net_wm_name;

    GRegex          *window_regex_comp;

    gboolean         criteria_updated;

    /* embed settings */
    gchar           *proc_name;
    gchar           *window_regex;
    gchar           *window_class;
    gchar           *launch_cmd;
    gchar           *label_fmt;
    gchar           *label_font;
    gint             poll_delay;
    gint             min_size;
    gboolean         expand;
    gboolean         show_handle;
}
EmbedPlugin;

/* Special values for EmbedPlugin::min_size */
#define EMBED_MIN_SIZE_MATCH_WINDOW 0

/* Special contents of launch_cmd */
#define EMBED_LAUNCH_CMD_SOCKET "%s"

/* Special contents of label_fmt */
#define EMBED_LABEL_FMT_TITLE "%t"

void
embed_search_again (EmbedPlugin *embed);
void
embed_stop_search (EmbedPlugin *embed);
void
embed_update_label (EmbedPlugin *embed);
void
embed_update_label_font (EmbedPlugin *embed);
void
embed_size_changed_simple (EmbedPlugin *embed);
void
embed_save (XfcePanelPlugin *plugin, EmbedPlugin *embed);

G_END_DECLS

#endif /* !__EMBED_H__ */


panel-plugin/embed-dialogs.h
/*  xfce4-embed-plugin - Embed arbitrary windows into the Xfce panel
 *
 *  Copyright (C) 2012 David Schneider <dnschneid@gmail.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef __EMBED_DIALOGS_H__
#define __EMBED_DIALOGS_H__

G_BEGIN_DECLS

void
embed_configure (XfcePanelPlugin *plugin, EmbedPlugin *embed);

void
embed_about (XfcePanelPlugin *plugin);

G_END_DECLS

#endif


panel-plugin/embed-dialogs.c
/*  xfce4-embed-plugin - Embed arbitrary windows into the Xfce panel
 *
 *  Copyright (C) 2012 David Schneider <dnschneid@gmail.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <string.h>
#include <gtk/gtk.h>
#include <X11/Xlib.h>

#include <libxfce4ui/libxfce4ui.h>
#include <libxfce4panel/xfce-panel-plugin.h>

#include "embed.h"
#include "embed-dialogs.h"

/* the website url */
#define PLUGIN_WEBSITE \
  "http://goodies.xfce.org/projects/panel-plugins/xfce4-embed-plugin"



static void
embed_configure_response (GtkWidget *dialog, gint response, EmbedPlugin *embed)
{
  gboolean result;

  if (response == GTK_RESPONSE_HELP) {
    /* show help */
    result = g_spawn_command_line_async (
        "exo-open --launch WebBrowser " PLUGIN_WEBSITE, NULL);

    if (G_UNLIKELY (result == FALSE))
      g_warning (_("Unable to open the following url: %s"), PLUGIN_WEBSITE);
  } else {
    /* remove the dialog data from the plugin */
    g_object_set_data (G_OBJECT (embed->plugin), "dialog", NULL);

    /* unlock the panel menu */
    xfce_panel_plugin_unblock_menu (embed->plugin);

    /* save the plugin */
    embed_save (embed->plugin, embed);

    /* destroy the properties dialog */
    gtk_widget_destroy (dialog);

    /* start a new search if necessary */
    if (embed->criteria_updated)
      embed_search_again (embed);
  }
}



static void
embed_proc_name_changed (GtkEditable *edit, EmbedPlugin *embed)
{
  g_free (embed->proc_name);
  embed->proc_name = g_strdup (gtk_entry_get_text (GTK_ENTRY (edit)));
  embed->criteria_updated = TRUE;
}



static void
embed_window_class_changed (GtkEditable *edit, EmbedPlugin *embed)
{
  g_free (embed->window_class);
  embed->window_class = g_strdup (gtk_entry_get_text (GTK_ENTRY (edit)));
  embed->criteria_updated = TRUE;
}



static void
embed_entry_set_good (GtkEntry *edit, gboolean good)
{
  if (good) {
    gtk_entry_set_icon_from_stock (edit, GTK_ENTRY_ICON_SECONDARY,
                                   GTK_STOCK_YES);
    gtk_entry_set_icon_tooltip_text (edit, GTK_ENTRY_ICON_SECONDARY,
                                     _("Input is valid"));
  } else {
    gtk_entry_set_icon_from_stock (edit, GTK_ENTRY_ICON_SECONDARY,
                                   GTK_STOCK_NO);
    gtk_entry_set_icon_tooltip_text (edit, GTK_ENTRY_ICON_SECONDARY,
                                     _("Input is invalid"));
  }
}



static void
embed_window_regex_changed (GtkEditable *edit, EmbedPlugin *embed)
{
  const gchar *text;
  GRegex *regex_comp;

  /* Confirm that the REGEX is okay before saving. */
  text = gtk_entry_get_text (GTK_ENTRY (edit));
  regex_comp = g_regex_new (text, G_REGEX_OPTIMIZE, 0, NULL);
  if (regex_comp) {
    g_free (embed->window_regex);
    if (embed->window_regex_comp)
      g_regex_unref (embed->window_regex_comp);
    embed->window_regex = g_strdup (text);
    embed->window_regex_comp = regex_comp;
    embed->criteria_updated = TRUE;
    embed_entry_set_good (GTK_ENTRY (edit), TRUE);
  } else {
    embed_entry_set_good (GTK_ENTRY (edit), FALSE);
  }
}



static void
embed_launch_cmd_changed (GtkEditable *edit, EmbedPlugin *embed)
{
  const gchar *text;
  gint argc;
  gchar **argv;

  /* Confirm that the command line is okay before saving. */
  text = gtk_entry_get_text (GTK_ENTRY (edit));
  if (*text) {
    if (!g_shell_parse_argv (text, &argc, &argv, NULL)) {
      embed_entry_set_good (GTK_ENTRY (edit), FALSE);
      return;
    }
    g_strfreev (argv);
  }

  g_free (embed->launch_cmd);
  embed->launch_cmd = g_strdup (text);
  embed->criteria_updated = TRUE;
  embed_entry_set_good (GTK_ENTRY (edit), TRUE);
}



static void
embed_label_fmt_changed (GtkEditable *edit, EmbedPlugin *embed)
{
  g_free (embed->label_fmt);
  embed->label_fmt = g_strdup (gtk_entry_get_text (GTK_ENTRY (edit)));
  embed_update_label (embed);
}



static void
embed_label_font_changed (GtkFontButton *font_button, EmbedPlugin *embed)
{
  g_free (embed->label_font);
  embed->label_font = g_strdup (gtk_font_button_get_font_name (font_button));
  embed_update_label_font (embed);
}



static void
embed_min_size_changed (GtkSpinButton *spin, EmbedPlugin *embed)
{
  embed->min_size = gtk_spin_button_get_value_as_int (spin);
  embed_size_changed_simple (embed);
}



static void
embed_expand_toggled (GtkToggleButton *toggle, EmbedPlugin *embed)
{
  embed->expand = gtk_toggle_button_get_active (toggle);
  xfce_panel_plugin_set_expand (embed->plugin, embed->expand);
}



static void
embed_handle_toggled (GtkToggleButton *toggle, EmbedPlugin *embed)
{
  embed->show_handle = gtk_toggle_button_get_active (toggle);
  if (embed->show_handle)
    gtk_widget_show (embed->handle);
  else
    gtk_widget_hide (embed->handle);
}



static void
set_tooltips (GtkWidget *widgetA, GtkWidget *widgetB, const gchar *text)
{
  gtk_widget_set_tooltip_text (widgetA, text);
  gtk_widget_set_tooltip_text (widgetB, text);
}



static GtkWidget *
add_frame (GtkWidget *content, gint rows, const gchar *title)
{
  GtkWidget *table, *frame;
  table = gtk_table_new (rows, 2, FALSE);
  gtk_table_set_col_spacings (GTK_TABLE (table), 12);
  gtk_table_set_row_spacings (GTK_TABLE (table), 6);
  frame = xfce_gtk_frame_box_new_with_content (title, table);
  gtk_box_pack_start_defaults (GTK_BOX (content), frame);
  return table;
}




static GtkWidget *
add_label (GtkWidget *table, gint row,
           GtkWidget *mnemonic_widget, const gchar *text)
{
  GtkWidget *label = gtk_label_new_with_mnemonic (text);
  if (mnemonic_widget) {
    gtk_misc_set_alignment (GTK_MISC (label), 1, 0.5f);
    gtk_label_set_mnemonic_widget (GTK_LABEL (label), mnemonic_widget);
  }
  gtk_table_attach_defaults (GTK_TABLE (table), label,
      0, mnemonic_widget ? 1 : 2, row, row+1);
  return label;
}



static void
add_entry (EmbedPlugin *embed, GtkWidget *table, gint row,
           const gchar *value, gboolean show_icon, gpointer callback,
           const gchar *labeltext, const gchar *tooltiptext)
{
  GtkWidget *label, *entry;
  entry = gtk_entry_new ();
  label = add_label (table, row, entry, labeltext);
  if (value) gtk_entry_set_text (GTK_ENTRY (entry), value);
  if (show_icon) embed_entry_set_good (GTK_ENTRY (entry), TRUE);
  g_signal_connect (G_OBJECT (entry), "changed", G_CALLBACK (callback), embed);
  set_tooltips (label, entry, tooltiptext);
  gtk_table_attach_defaults (GTK_TABLE (table), entry, 1, 2, row, row+1);
}



static void
add_fontbutton (EmbedPlugin *embed, GtkWidget *table, gint row,
                const gchar *value, gpointer callback,
                const gchar *labeltext, const gchar *tooltiptext)
{
  GtkWidget *label, *button;
  button = gtk_font_button_new ();
  label = add_label (table, row, button, labeltext);
  if (value) gtk_font_button_set_font_name (GTK_FONT_BUTTON (button), value);
  g_signal_connect (G_OBJECT (button), "font-set",
                    G_CALLBACK (callback), embed);
  set_tooltips (label, button, tooltiptext);
  gtk_table_attach_defaults (GTK_TABLE (table), button, 1, 2, row, row+1);
}



static void
add_spinbutton (EmbedPlugin *embed, GtkWidget *table, gint row,
                gint value, gpointer callback,
                const gchar *labeltext, const gchar *tooltiptext)
{
  GtkWidget *label, *button;
  button = gtk_spin_button_new_with_range (0, G_MAXINT, 1);
  label = add_label (table, row, button, labeltext);
  gtk_spin_button_set_value (GTK_SPIN_BUTTON (button), value);
  g_signal_connect (G_OBJECT (button), "value-changed",
                    G_CALLBACK (callback), embed);
  set_tooltips (label, button, tooltiptext);
  gtk_table_attach_defaults (GTK_TABLE (table), button, 1, 2, row, row+1);
}



static void
add_checkbutton (EmbedPlugin *embed, GtkWidget *table, gint row,
                 gboolean value, gpointer callback,
                 const gchar *labeltext, const gchar *tooltiptext)
{
  GtkWidget *button = gtk_check_button_new_with_mnemonic (labeltext);
  gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (button), value);
  g_signal_connect (G_OBJECT (button), "toggled", G_CALLBACK (callback), embed);
  gtk_widget_set_tooltip_text (button, tooltiptext);
  gtk_table_attach_defaults (GTK_TABLE (table), button, 1, 2, row, row+1);
}



void
embed_configure (XfcePanelPlugin *plugin, EmbedPlugin *embed)
{
  GtkWidget *dialog, *content, *table;
  gchar *text;

  /* block the plugin menu */
  xfce_panel_plugin_block_menu (plugin);

  /* stop searches */
  embed_stop_search (embed);

  /* create the dialog */
  dialog = xfce_titled_dialog_new_with_buttons (
      _("Embed Plugin"),
      GTK_WINDOW (gtk_widget_get_toplevel (GTK_WIDGET (plugin))),
      GTK_DIALOG_DESTROY_WITH_PARENT | GTK_DIALOG_NO_SEPARATOR,
      GTK_STOCK_HELP, GTK_RESPONSE_HELP,
      GTK_STOCK_CLOSE, GTK_RESPONSE_OK,
      NULL);

  content = gtk_dialog_get_content_area (GTK_DIALOG (dialog));

  table = add_frame (content, 2, _("Application Launching"));
  text = g_strdup_printf (
      _("If a window is not found (or there are no criteria), a command can\n"
        "optionally be launched. The command can either result in a window\n"
        "that matches the below criteria, or it can use the socket ID passed\n"
        "to it (%s) to embed itself automatically."), EMBED_LAUNCH_CMD_SOCKET);
  add_label (table, 0, NULL, text);
  g_free (text);
  /* launch_cmd */
  text = g_strdup_printf (
      _("Leave blank to not launch anything\n"
        "%s expands to the socket ID"), EMBED_LAUNCH_CMD_SOCKET);
  add_entry (embed, table, 1, embed->launch_cmd, TRUE, embed_launch_cmd_changed,
           _("L_aunch command"), text);
  g_free (text);

  /* poll_delay */
  /* No UI element. Generally polling is unnecessary, unless you have a very
   * strange window that you're trying to match that is not uniquely
   * identifiable when it is mapped. */


  table = add_frame (content, 4, _("Selection Criteria"));
  add_label (table, 0, NULL,
      _("The window to embed must match all of the non-blank criteria.\n"
        "Leave everything blank to rely on a launch command with socket ID."));
  /* proc_name */
  add_entry (embed, table, 1, embed->proc_name, FALSE, embed_proc_name_changed,
           _("_Process name"),
           _("Match the window's application's process name\n"
             "Leave blank if it is not a criterion"));

  /* window_class */
  add_entry (embed, table, 2, embed->window_class, FALSE,
             embed_window_class_changed,
           _("_Window class"),
           _("Match the window's class\n"
             "Leave blank if it is not a criterion"));

  /* window_regex */
  add_entry (embed, table, 3, embed->window_regex, TRUE,
             embed_window_regex_changed,
           _("Window _title"),
           _("Match the window's title using a REGEX\n"
             "Leave blank if it is not a criterion"));


  table = add_frame (content, 3, _("Display"));
  /* label_fmt */
  text = g_strdup_printf (
      _("Leave blank to hide the label\n"
        "%s expands to the embedded window's title"), EMBED_LABEL_FMT_TITLE);
  add_entry (embed, table, 0, embed->label_fmt, FALSE, embed_label_fmt_changed,
           _("_Label format"), text);
  g_free (text);

  /* label font */
  add_fontbutton (embed, table, 1, embed->label_font, embed_label_font_changed,
                _("Label _font"),
                _("Choose the label font"));

  /* min_size */
  add_spinbutton (embed, table, 2, embed->min_size, embed_min_size_changed,
                _("Minimum _size (px)"),
                _("Minimum size of the embedded window\n"
                  "Set to 0 to keep the original window size"));

  /* expand */
  add_checkbutton (embed, table, 3, embed->expand, embed_expand_toggled,
                 _("_Expand"),
                 _("Use up all available panel space"));

  /* show_handle */
  add_checkbutton (embed, table, 4, embed->show_handle, embed_handle_toggled,
                 _("Show _handle"),
                 _("Display a handle at the top of the plugin"));


  /* center dialog on the screen */
  gtk_window_set_position (GTK_WINDOW (dialog), GTK_WIN_POS_CENTER);

  /* set dialog icon */
  gtk_window_set_icon_name (GTK_WINDOW (dialog), "xfce4-settings");

  /* link the dialog to the plugin, so we can destroy it when the plugin
   * is closed, but the dialog is still open */
  g_object_set_data (G_OBJECT (plugin), "dialog", dialog);

  /* criteria have not been updated */
  embed->criteria_updated = FALSE;

  /* connect the reponse signal to the dialog */
  g_signal_connect (G_OBJECT (dialog), "response",
                    G_CALLBACK (embed_configure_response), embed);

  /* show the entire dialog */
  gtk_widget_show_all (dialog);
}
